Change behavior of auto properties to not override explicit configuration

thomaslevesque opened the pull request:
This PR is an experiment to demonstrate a new behavior for "auto properties", as discussed in the chat.

Fixes #962

blairconrad commented on src/FakeItEasy/Core/FakeManager.PropertyBehaviorRule.cs:
```diff
@@ -39,25 +39,18 @@ public static bool IsPropertyGetter(MethodInfo method)
                 return method.IsSpecialName && method.Name.StartsWith("get_", StringComparison.Ordinal);
             }
 
+            public bool IsMatchForSetter(IFakeObjectCall fakeObjectCall) => IsPropertySetter(fakeObjectCall);
```
is the `this` still required (encouraged)?

blairconrad replied:
I'm not really sure why we need this method as well as `IsPropertySetter`

thomaslevesque replied:
> is the this still required (encouraged)?

Yes. Not sure why StyleCop doesn't detect it. Perhaps it's confused by the expression-bodied method.

> I'm not really sure why we need this method as well as IsPropertySetter

We don't really need it, but the `IsPropertySetter` instance method is private, and if I made it public, it would be strange if `IsPropertyGetter` were not also public, although it's not needed. Also, I thought `IsMatchForSetter` communicated the intent better.

thomaslevesque replied:
But I can change it if you prefer.

blairconrad replied:
I guess `IsMatchForSetter` makes it sound to me like we're comparing the rule against a particular setter, but perhaps that's just my brain making assumptions.
This is not such a big deal to me. I appreciate the the explanation.

thomaslevesque replied:
If you have a better name to suggest, I'll gladly take it. I didn't think too long about this one.

blairconrad replied:
I don't think it's worth renaming.
The only alternative     I can think of that I like is `IsPropertySetter`. :wink:

blairconrad commented on src/FakeItEasy/Core/FakeManager.PropertySetterRule.cs:
```diff
@@ -32,17 +32,31 @@ public void Apply(IInterceptedFakeObjectCall fakeObjectCall)
+                    .Reverse()
+                    .Select(metadata => metadata.Rule)
+                    .OfType<PropertyBehaviorRule>()
+                    .FirstOrDefault(rule => rule.IsMatchForSetter(fakeObjectCall));
```
it took me a while to follow all the reverses and adding to last. I believe it's correct, but find it hard to follow.

thomaslevesque replied:
Yes, it's just that I wanted to search from the end. But now that I think about it, I could use `LastOrDefault` instead.

blairconrad replied:
Oh, yes. `LastOrDefault` and no `Reverse` would be better. And perhaps a comment at the insertion site to indicate why we're shoving the rule at the end instead of the beginning like all the other rules?

thomaslevesque replied:
Sure, I'll add an explanation in the comments.

thomaslevesque commented:
I think I made all the requested changes. I'll squash everything before merging.

blairconrad commented on src/FakeItEasy/Core/FakeManager.PropertySetterRule.cs:
```diff
@@ -32,17 +32,34 @@ public void Apply(IInterceptedFakeObjectCall fakeObjectCall)
-                this.fakeManager.AllUserRules.AddFirst(newRule);
+                //// Setting the property adds a PropertyBehaviorRule to store the assigned value
+                //// The PropertyBehaviorRule is added at the end of user rules to avoid overriding
+                //// explicit configurations of the property made with CallTo.
```
Such a nitpick, but the `////` looks funny to me. This is the format that StyleCop suggest for commenting out code. How do you feel about just `//`?

(And while I'm here, I think you forgot a "." at the end of the first line.)

adamralph replied:
I guess StyleCop is complaining about the blank line after the comment when it only has `//`.

What does the comment apply to? The next line of code or the whole method? (This is exactly the ambiguity that StyleCop does't like.) If it's the next line, then the line break should be removed. If it's the whole method, then it should move out to above the method signature.

thomaslevesque replied:
The comment applies to the next statement (the LINQ query).

I'm a bit confused... I *think* I had introduced the blank line and the `////` to make StyleCop happy, but there's no warning with `//` and no blank line, so I'll do just that.

Also, it seems I had missed @blairconrad's comment...

thomaslevesque commented:
> In #962 (comment), you'd mentioned that you were going to add documentation. Was that the docs in the code, or did you mean something to go on readthedocs to let users know about the new behaviour?

Oops! No, I intended to update the docs, not just add comments to the code.

thomaslevesque commented:
I updated the docs. I'm a bit uneasy about what I wrote, though, because it's not strictly true... Setting a property whose getter has been explicitly configured actually *does* something, but the effect won't be visible while the getter configuration is active (see the second snippet in https://github.com/FakeItEasy/FakeItEasy/issues/962#issuecomment-270382936). But I'm afraid that trying to explain this will confuse the reader even more... What do you think?

thomaslevesque commented on docs/read-write-property-behavior.md:
````diff
@@ -14,11 +14,28 @@ var fakeShop = A.Fake<ICandyShop>();
 ```
 
-This behaviour can be used to 
+This behavior can be used to
````
For consistency with the rest of the documents.

We have 11 occurrences of "behaviour" (all in the docs), and 271 occurrences of "behavior" (35 of them in the docs)...

blairconrad commented:
Sorry for the delay, @thomaslevesque. Truth is, I'd forgotten about your update. 😊
Nice catch on the "behaviour".

I understand your concern about lying in the docs, but sometimes it's better to lie about what's happening internally if it better explains the view from the outside.
I'd support merging the text you have, but want to present an alternative idea.
We could, instead of appending your explanation, at the top of the topic indicate that the behavio(u)r we describe applies only to properties that are not otherwise configured as in [Specifying Return Values](http://fakeiteasy.readthedocs.io/en/stable/specifying-return-values/), [Throwing Exceptions](http://fakeiteasy.readthedocs.io/en/stable/throwing-exceptions/) , etc., as you've indicated in your bold note.
Does that approach appeal at all?
Of course if it did, the topic would probably be better named "Default Read/Write Property Behavior". Or we could even roll it into [Default fake behavior](http://fakeiteasy.readthedocs.io/en/stable/default-fake-behavior/).

thomaslevesque commented:
@blairconrad I like your idea, I'll give it a try.

thomaslevesque commented:
> Or we could even roll it into Default fake behavior.

I think I'd rather keep it separate, but we can add a link to it in Default fake behavior.

blairconrad commented:
> I think I'd rather keep it separate, but we can add a link to it in Default fake behavior.

Yes. This is better.

thomaslevesque commented:
OK, I tried to apply your idea, but I'm not sure I like it better. I think it's useful to have an example of what happens when the default behavior no longer applies, but that example has to come *after* we explained the default behavior.

blairconrad commented:
Thanks, @thomaslevesque. I agree with you. As it is, I don't like it better either. If we're going to document "default read/write property behavior", I think it belongs in the docs nearer (or maybe "in" is a better approach after all) the [Default fake behavior](http://fakeiteasy.readthedocs.io/en/stable/default-fake-behavior/) topic, which supports your idea of explaining the default behavior before explaining what happens when we explicitly configure the property.

thomaslevesque commented:
> If we're going to document "default read/write property behavior", I think it belongs in the docs nearer (or maybe "in" is a better approach after all) the Default fake behavior topic, which supports your idea of explaining the default behavior before explaining what happens when we explicitly configure the property.

I'm not sure I understand; what are you suggesting?

blairconrad commented:
> I'm not sure I understand; what are you suggesting?

Move the topic for the default read/write property behaviour into [Default fake behavior](http://fakeiteasy.readthedocs.io/en/stable/default-fake-behavior/).
(Alternatively, keep the topic separate, but move it up next to "Default fake behavior" in the table of contents, but I think inlining it is better.)

Then, as you suggested, people see the default behaviour first, and then later on in the docs we talk about how to override the getter and setter with fancier behaviour.
I quick roughed something up for you to get the sense of it, in case I'm still not being clear: https://github.com/blairconrad/FakeItEasy/commit/40070b0c7f2e029a69772ef77154debf27490475

(Also, I accidentally pushed a commit to _your_ branch using that scary new GitHub "push to other people's branches" technology, but was able to reset the head of the branch and repush. Which is even scarier! I hope I didn't hurt anything.)

thomaslevesque commented:
Ah, I see. I wasn't completely awake when I posted this morning... Your changes look pretty good!

> (Also, I accidentally pushed a commit to your branch using that scary new GitHub "push to other people's branches" technology, but was able to reset the head of the branch and repush. Which is even scarier! I hope I didn't hurt anything.)

Oh, I don't mind you pushing to my branch. In fact, I think I'm going to merge your branch into mine, which will have the same result :wink:

thomaslevesque commented:
I took your commit, fixed the line endings, added back `read-write-property-behavior.md`, squashed the implementation commits together, and squashed the documentation commits together (attributed to you, since none of what I had done remained ;))

blairconrad commented:
>Oh, I don't mind you pushing to my branch.

Generous of you. I feel likewise.
My main concern was how easy it was for me to _accidentally_ do it!
In retrospect it makes sense. I checked out your branch and tracked it, so when I pushed without specifying a remote…

> added back `read-write-property-behavior.md`

Thanks. I'm surprised the line endings changed. I'd told emacs to visit the file "literally", which I think is supposed to leave _all_ the bytes alone. Ah well.

Also thanks for the attribution, not that it matters. It's all about the result!

blairconrad commented:
It looks good to me, @thomaslevesque. Thanks for the work.
Leaving for a bit again in case @adamralph wants to weigh in on the changed behavio(u)r.

thomaslevesque commented:
@adamralph, you're right, there should be more tests to cover the new behavior. I'll add them.

> at the black box level

What do you mean by that? specs?

adamralph commented:
@thomaslevesque yep, specs is what I meant. Would that be appropriate for this?

thomaslevesque commented:
> Would that be appropriate for this?

I think so

thomaslevesque commented:
I added some specs. The read/write property behavior was more or less covered in several spec classes, but none of them was specifically testing it, so I introduced a new spec class.

thomaslevesque commented:
I edited the comment in `PropertySetterRule` and fixed some new FxCop warnings (amended the commit).

adamralph commented:
Thanks @thomaslevesque !

thomaslevesque commented:
Thanks for the merge
